fem/hybridization.cpp

```diff
@@ -821,7 +821,15 @@ void Hybridization::ReduceRHS(const Vector &b, Vector &b_r) const
    else
    {
       Vector bl(pC ? pC->Height() : Ct->Width());
-      pC ? pC->Mult(bf, bl) : Ct->MultTranspose(bf, bl);
+      if (pC)
+      {
+         pC->Mult(bf, bl);
+      }
+      else
+      {
+         if (Device::IsEnabled()) { Ct->BuildTranspose(); }
```

Comment:
Why not just build the transpose always?

Reply:
The transpose is not going to be used on cpu, so this saves the memory and the time.

Reply:
Won't the transpose still be used, e.g. https://github.com/mfem/mfem/blob/256bb5627fade2380ca51d08b7b908d8fe71feaa/linalg/sparsemat.cpp#L845-L848
but I see what you mean that it is not strictly necessary on CPU.

Is the memory and time significant (i.e. worth slightly complicating the code)?

Reply:
Actually, on cpu the transpose will also be used, if it is built, but that is not needed and we can save the time for constructing the transpose and the memory for storing it.

Reply:
It also makes me wonder if this is not an idiom that will become increasingly common, maybe something like `SparseMatrix::EnsureDeviceTranspose` would be helpful.

It might make sense to have a way of communicating "this is a matrix and I want to be able to multiply by its transpose" without worrying about details of the device, or weighing efficiency tradeoffs.

Reply:
> Is the memory and time significant (i.e. worth slightly complicating the code)?

Constructing the transpose matrix essentially doubles the memory cost, so I think it's worth it.

Reply:
Adding `SparseMatrix::EnsureDeviceTranspose` sounds good to me. It will conditionally call `BuildTranspose` based on the `mfem::Device` configuration. Instead of just checking `Device::IsEnabled` (which checks if any backend other than `Backend::CPU` is enabled) we can be a bit more precise and check if any non-serial-CPU backend is enabled, e.g. with `Device::Allows(~Backend::CPU_MASK)`.

With that addition, we should replace the other conditional calls to `BuildTranspose` with `EnsureDeviceTranspose`. Also, we'll need to update the `MFEM_VERIFY` in `SparseMatrix::AddMultTranspose`.

Finally, since `EnsureDeviceTranspose` should construct the transpose for `OpenMP` backends too, it is probably better to call it `EnsureTranspose`.

Reply:
Let's do it here then, if that works for you @tomstitt.

Reply:
OK, I just added it. Let me know if the latest commit looks good to both of you.

Reply:
Looks great! Thanks @pazner